doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1149 @@
+
+Some of these difficulties arise due to the fact that finding the correct shape
+of the output array and finding the correct datatype are closely related.
+
+**Implementation:**
+
+There are two distinct cases given above: First, when the user has provided no
+dtype information, and second when the user provided a DType class – 
+a notion that is currently represented e.g. by ``"S"``, but not cleanly supported.
+
+In the first case, it is necessary to establish a mapping from the python type
+that was put in to the DType class.
+Further, the correct dtype instance needs to be found.
+This shall be implemented by the use of two informations:
+
+1. ``type``: The current type attribute to indicate which python type is
+   associated with the DType class (this is a *class* attribute that always
+   exist for any datatype and is not limited to array coercion).
+2. The reverse lookup remain hardcoded for the basic Python types initially.
+   Otherwise rely on the ``type`` attribute, and at least initially may
+   inforce deriving the scalar from a NumPy provided scalar.
+   This method may be expanded later (see alternatives).
+3. ``__discover_descr_from_pyobject__(cls, obj) -> dtype``: A classmethod that
+   returns the correct descriptor given the input object.
+   *Note that only parametric DTypes have to implement this*, most datatypes
+   can simply use a default (singleton) dtype instance which is found only
+   based on the ``type(obj)`` of the Python object.
+
+The ``type`` which is already associated with any dtype through the ``dtype.type``
+attribute maps the dtype to the Python type.
+However, initially we anticipate automatic conversion only for scalars which
+subclass from a base scalar object, such as ``np.generic`` (although possibly
+a class below ``np.generic`` to avoid arry-scalar behaviour). 
+This will be cached globally to create a mapping (dictionary)
+``knwon_python_types[type] = DType``.
+NumPy currently uses a small hard-coded mapping and conversion of numpy scalars
+(inheriting from ``np.generic``) to achieve this, however, this forces a certain
+structure on the associated ``type``.
+
+.. note::
+
+    Python integers do not have a clear/specific NumPy type associated to them
+    right now.  Instead they will need to be be implemented using an
+    ``AbstractPyInt``, this DType class, can than provide
+    ``__discover_descr_from_pyobject__`` and return the actual dtype which
+    is e.g. ``np.dtype("int64")``.
+    For dispatching/promotion in ufuncs, it will also probably be necessary
+    to dynamically create ``AbstractPyInt[value]`` classes (creation can be
+    cached), so that they can provide the current functionality provided by
+    ``np.result_type(python_integer, array)``.
+
+To allow for a DType to accept specific inputs as known scalars, we will
+initially use a method such as ``known_scalar_type`` or ``known_scalar_object``
+which allows discovering a e.g. ``vector`` as a scalar instead of an instance
```

Comment:
```suggestion
which allows discovering a e.g. ``vector`` as a scalar instead of an instance
```
Instead of "instance" do you mean to say something like "collection of scalars"?